contracts/0.8.9/oracle/ValidatorsExitBusOracle.sol

```diff
@@ -0,0 +1,507 @@
+            requestsCount: data.requestsCount.toUint64(),
+            requestsProcessed: data.requestsCount.toUint64(),
+            dataFormat: uint16(DATA_FORMAT_LIST)
+        });
+
+        if (data.requestsCount == 0) {
+            return;
+        }
+
+        RateLimit
+            .load(RATE_LIMIT_POSITION)
+            .recordUsageAt(data.refSlot, data.requestsCount * 10**18)
+            .store(RATE_LIMIT_POSITION);
+
+        TOTAL_REQUESTS_PROCESSED_POSITION.setStorageUint256(
+            TOTAL_REQUESTS_PROCESSED_POSITION.getStorageUint256() + data.requestsCount
+        );
+    }
+
+    function _processExitRequestsList(bytes calldata data) internal returns (uint256) {
+        uint256 offset;
+        uint256 offsetPastEnd;
+        assembly {
+            offset := data.offset
+            offsetPastEnd := add(offset, data.length)
+        }
+
+        mapping(uint256 => RequestedValidator) storage _lastReqValidatorIndices =
+            _storageLastRequestedValidatorIndices();
+
+        uint256 lastDataWithoutPubkey = 0;
+        uint256 lastNodeOpKey = 0;
+        uint256 lastValIndex;
+        bytes calldata pubkey;
+
+        assembly {
+            pubkey.length := 48
+        }
+
+        while (offset < offsetPastEnd) {
+            uint256 dataWithoutPubkey;
+            assembly {
+                // 16 most significant bytes are taken by module id, node op id, and val index
+                dataWithoutPubkey := shr(128, calldataload(offset))
+                // the next 48 bytes are taken by the pubkey
+                pubkey.offset := add(offset, 16)
+                // totalling to 64 bytes
+                offset := add(offset, 64)
+            }
+            //                              dataWithoutPubkey
+            // MSB <---------------------------------------------------------------------- LSB
+            // | 128 bits: zeros | 24 bits: moduleId | 40 bits: nodeOpId | 64 bits: valIndex |
+            //
+            if (dataWithoutPubkey <= lastDataWithoutPubkey) {
+                revert InvalidRequestsDataSortOrder();
+            }
+
+            uint256 valIndex = uint64(dataWithoutPubkey);
+            uint256 nodeOpId = uint40(dataWithoutPubkey >> 64);
+            uint256 moduleId = uint24(dataWithoutPubkey >> (64 + 40));
+
+            if (moduleId == 0) {
+                revert InvalidRequestsData();
+            }
+
+            uint256 nodeOpKey = _computeNodeOpKey(moduleId, nodeOpId);
+            if (nodeOpKey != lastNodeOpKey) {
+                if (lastNodeOpKey != 0) {
+                    _lastReqValidatorIndices[lastNodeOpKey] =
+                        RequestedValidator(true, uint64(lastValIndex));
+                }
+                lastNodeOpKey = nodeOpKey;
+            }
+
+            lastValIndex = valIndex;
+            lastDataWithoutPubkey = dataWithoutPubkey;
+
+            emit ValidatorExitRequest(moduleId, nodeOpId, valIndex, pubkey, block.timestamp);
```

Comment:
`emit ValidatorExitRequest(moduleId, nodeOpId, valIndex, pubkey, _getTime());` to work properly in mocks